File: app/src/main/java/com/totemsoft/screenlookcount/ActivityMain.java
package com.totemsoft.screenlookcount;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;
import com.totemsoft.screenlookcount.background.UpdateService;
import com.totemsoft.screenlookcount.db.LookCountDbManager;
import com.totemsoft.screenlookcount.db.OnDatabaseListener;
import com.totemsoft.screenlookcount.fragment.FragmentAbout;
import com.totemsoft.screenlookcount.fragment.FragmentMain;
import com.totemsoft.screenlookcount.utils.C;
import com.totemsoft.screenlookcount.utils.Utils;

import java.lang.ref.WeakReference;
import java.util.Calendar;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * This is the main activity which acts as a holder for {@link FragmentMain} and {@link FragmentAbout} fragments.
 * Handles Toolbar options as well.
 *
 * @author antonina.tkachuk
 */
public class ActivityMain extends AppCompatActivity {
    @Bind(R.id.sv_container)
    ScrollView fragmentContainer;
    private Tracker tracker;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        Utils.addFragment(this, FragmentMain.class, C.FRAGMENT_TAG_MAIN, fragmentContainer, false, null);
        saveFirstRunDate();
        tracker = ((AnalyticsApplication) getApplication()).getDefaultTracker();
    }

    @Override
    protected void onStart() {
        super.onStart();
        // start the service and let it work ever after (with a "sticky" flag)
        startService(new Intent(this, UpdateService.class));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_clear_data:
                clearDataAction();
                return true;
            case R.id.action_share:
                shareWithFriendsAction();
                return true;
            case R.id.action_rate_app:
                rateTheAppActionAction();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @OnClick(R.id.b_about)
    public void showAbout() {
        if (Utils.isFragmentVisible(this, C.FRAGMENT_TAG_MAIN)) {
            Utils.addFragment(this, FragmentAbout.class, C.FRAGMENT_TAG_ABOUT, fragmentContainer.getId(), true, null);
        } else {
            getSupportFragmentManager().popBackStack();
        }
    }

    private void saveFirstRunDate() {
        if (Utils.getPreference(this, C.PREFERENCE_FIRST_RUN_DATE) == 0) {
            Utils.savePreference(this, C.PREFERENCE_FIRST_RUN_DATE, Calendar.getInstance().getTimeInMillis());
        }
    }

    private void showClearAllDataUndoSnackbar() {
        Snackbar undoSnackbar = Snackbar.make(fragmentContainer, getString(R.string.snack_bar_clear_text), Snackbar.LENGTH_LONG)
                .setAction(getString(R.string.snack_bar_clear_action_undo), new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        LookCountDbManager.getManager(ActivityMain.this).setDbClearedFlag(false);
                        updateMainView();
                        tracker.send(new HitBuilders.EventBuilder()
                                .setCategory("Buttons")
                                .setAction("UNDO")
                                .build());
                    }
                })
                .setCallback(new Snackbar.Callback() {
                    @Override
                    public void onDismissed(Snackbar snackbar, int event) {
                        super.onDismissed(snackbar, event);
                        if (event != Snackbar.Callback.DISMISS_EVENT_ACTION) {
                            clearAllData();
                        }
                    }
                });

        // styling
        undoSnackbar.setActionTextColor(ContextCompat.getColor(this, R.color.snack_bar_action));
        View snackbarView = undoSnackbar.getView();
        int snackbarTextId = android.support.design.R.id.snackbar_text;
        TextView textView = (TextView) snackbarView.findViewById(snackbarTextId);
        textView.setTextColor(ContextCompat.getColor(this, R.color.snack_bar_text));
        snackbarView.setBackgroundColor(ContextCompat.getColor(this, R.color.snack_bar_background));

        undoSnackbar.show();
    }

    private void clearAllData() {
        LookCountDbManager.getManager(this).setDbListener(new ClearDBListener(this));
        LookCountDbManager.getManager(this).dropDatabase();
    }

    private void updateMainView() {
        Fragment mainFragment = Utils.getFragmentByTag(this, C.FRAGMENT_TAG_MAIN);
        if (mainFragment instanceof FragmentMain) {
            ((FragmentMain) mainFragment).updateViewAfterDBOperation();
        }
    }

    private void clearDataAction() {
        LookCountDbManager.getManager(this).setDbClearedFlag(true);
        updateMainView();
        showClearAllDataUndoSnackbar();
        tracker.send(new HitBuilders.EventBuilder()
                .setCategory("Options")
                .setAction("Clear all Data")
                .build());
    }

    private void rateTheAppActionAction() {
        final String appPackageName = getPackageName();
        try {
            final Intent googlePlayIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.rate_google_play_uri, appPackageName)));
            startActivity(googlePlayIntent);
        } catch (ActivityNotFoundException exception) {
            final Intent googlePlayWebPageIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.rate_google_play_uri_browser, appPackageName)));
            startActivity(googlePlayWebPageIntent);
        }
        tracker.send(new HitBuilders.EventBuilder()
                .setCategory("Options")
                .setAction("Rate the app")
                .build());
    }

    private void shareWithFriendsAction() {
        Intent shareIntent = new Intent();
        shareIntent.setAction(Intent.ACTION_SEND);
        shareIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text, getPackageName()));
        shareIntent.setType("text/plain");
        startActivity(Intent.createChooser(shareIntent, getResources().getText(R.string.share_chooser)));
        tracker.send(new HitBuilders.EventBuilder()
                .setCategory("Options")
                .setAction("Share with Friends")
                .build());
    }

    private static final class ClearDBListener implements OnDatabaseListener {
        private final WeakReference context;

        private ClearDBListener(FragmentActivity context) {
            this.context = new WeakReference<>(context);
        }

        @Override
        public void onDatabaseDataCleared() {
            FragmentActivity activity = (FragmentActivity) this.context.get();
            if (activity != null && activity instanceof ActivityMain) {
                ((ActivityMain) activity).updateMainView();
            }
        }
    }
}

File: app/src/main/java/com/totemsoft/screenlookcount/ActivityMain.kt
package com.totemsoft.screenlookcount

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import com.totemsoft.screenlookcount.background.LookCounterService
import com.totemsoft.screenlookcount.db.ScreenCounterDb
import com.totemsoft.screenlookcount.fragment.about.FragmentAbout
import com.totemsoft.screenlookcount.fragment.main.FragmentMain
import com.totemsoft.screenlookcount.utils.*
import kotlinx.android.synthetic.main.activity_main.*

/**
 * This is the main activity which acts as a holder for [FragmentMain] and [FragmentAbout] fragments.
 * Handles Toolbar actions as well.
 *
 * @author Antonina
 */
class ActivityMain : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initView()
    }

    override fun onStart() {
        super.onStart()

        if (AppPreferences.shouldRunCountingService) {
            startService()
        }

        if (!isFragmentInBackStack(C.FRAGMENT_TAG_MAIN)) {
            addFragment(FragmentMain::class.java, C.FRAGMENT_TAG_MAIN, sv_container, false, null)
        }
    }

    private fun initView() {
        b_about.setOnClickListener {
            showOrHideAbout()
        }
    }

    private fun updateMainView() {
        val mainFragment = getFragmentByTag(C.FRAGMENT_TAG_MAIN)
        if (mainFragment is FragmentMain) {
            mainFragment.getMainPresenter().setCountersValues(this)
        }
    }

    private fun showOrHideAbout() {
        if (isFragmentVisible(C.FRAGMENT_TAG_MAIN)) {
            addFragment(FragmentAbout::class.java, C.FRAGMENT_TAG_ABOUT, sv_container.id, true, null)
        } else {
            supportFragmentManager.popBackStack()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)

        menu.findItem(R.id.action_switch_service).title =
                if (AppPreferences.shouldRunCountingService)
                    getString(R.string.action_stop_counting)
                else
                    getString(R.string.action_start_counting)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_switch_service -> {
                switchService(item)
                true
            }
            R.id.action_clear_data -> {
                clearDataAction()
                true
            }
            R.id.action_share -> {
                shareWithFriendsAction()
                true
            }
            R.id.action_rate_app -> {
                rateAppAction()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun startService() {
        startService(Intent(this, LookCounterService::class.java))
    }

    private fun stopService() {
        stopService(Intent(this, LookCounterService::class.java))
    }

    private fun clearAllData() {
        ScreenCounterDb.getDatabase(this).dropDb()
    }

    private fun showClearAllDataUndoSnackbar() {
        val undoSnackbar = Snackbar.make(sv_container, getString(R.string.snack_bar_clear_text), Snackbar.LENGTH_LONG)
                .setAction(getString(R.string.snack_bar_clear_action_undo)) {
                    ScreenCounterDb.getDatabase(this).setDbClearFlag(false)
                    updateMainView()
                }
                .addCallback(object : Snackbar.Callback() {
                    override fun onDismissed(snackbar: Snackbar?, event: Int) {
                        super.onDismissed(snackbar, event)
                        if (event != Snackbar.Callback.DISMISS_EVENT_ACTION) {
                            clearAllData()
                        }
                    }
                })

        // styling
        undoSnackbar.setActionTextColor(ContextCompat.getColor(this, R.color.snack_bar_action))
        val snackbarView = undoSnackbar.view
        val snackbarTextId = android.support.design.R.id.snackbar_text
        val textView = snackbarView.findViewById(snackbarTextId) as TextView
        textView.setTextColor(ContextCompat.getColor(this, R.color.snack_bar_text))
        snackbarView.setBackgroundColor(ContextCompat.getColor(this, R.color.snack_bar_background))

        undoSnackbar.show()
    }

    private fun switchService(menuItem: MenuItem) {
        val startCountingTitle = getString(R.string.action_start_counting)
        val stopCountingTitle = getString(R.string.action_stop_counting)
        if (menuItem.title == startCountingTitle) {
            AppPreferences.shouldRunCountingService = true
            startService()
            menuItem.title = stopCountingTitle
        } else {
            AppPreferences.shouldRunCountingService = false
            stopService()
            menuItem.title = startCountingTitle
        }
    }

    private fun clearDataAction() {
        ScreenCounterDb.getDatabase(this).setDbClearFlag(true)
        updateMainView()
        showClearAllDataUndoSnackbar()
    }

    private fun rateAppAction() {
        val appPackageName = packageName
        try {
            val googlePlayIntent = Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.rate_google_play_uri, appPackageName)))
            startActivity(googlePlayIntent)
        } catch (exception: ActivityNotFoundException) {
            val googlePlayWebPageIntent = Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.rate_google_play_uri_browser, appPackageName)))
            startActivity(googlePlayWebPageIntent)
        }
    }

    private fun shareWithFriendsAction() {
        val shareIntent = Intent()
        with(shareIntent) {
            action = Intent.ACTION_SEND
            putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text, packageName))
            type = "text/plain"
            startActivity(Intent.createChooser(this, resources.getText(R.string.share_chooser)))
        }
    }
}